### PR TITLE
iBVP dataset, iBVPNet model and MACC evaluation metrics

### PR DESCRIPTION
**Pull request by jnj256**
Dear rPPG-Toolbox team,

Below we summarize our contributions that have been published in our recent work on [iBVP dataset](https://www.mdpi.com/2079-9292/13/7/1334).

Through our work, we primarily contribute a novel rPPG dataset, which we refer to as 'iBVP dataset'. Additionally, we introduce iBVPNet model, which was found to be an effective 3D-CNN architecture, with results presented in the paper. We further discover the use of maximum amplitude of cross-correlation (MACC) as highly suitable evaluation metrics to validate rPPG methods.

Researchers will be able to access our dataset by referring to the details mentioned on this [GitHub page](https://github.com/PhysiologicAILab/iBVP-Dataset).

Our code builds upon the [rPPG-Toolbox](https://github.com/ubicomplab/rPPG-Toolbox) repo, and all the contributions from our side can be found [in this forked repo](https://github.com/PhysiologicAILab/rPPG-Toolbox). We would like to create a pull request to merge our contributions. For the same, we summarize the key changes below, and provide a link for detailed [difference log](https://github.com/ubicomplab/rPPG-Toolbox/compare/main...PhysiologicAILab:rPPG-Toolbox:main).

**Added:**
* dataset/data_loader/IBVPLoader.py: For loading RGB video frames of iBVP dataset
* neural_methods/model/iBVPNet.py: The iBVPNet model introduced in our work
* neural_methods/trainer/iBVPNetTrainer.py: Trainer for iBVPNet
* "train" and "infer" configs for iBVP dataset as well as iBVPNet model

**Updated:**
* dataset/data_loader/__init__.py: Added import statement for IBVPLoader
* neural_methods/trainer/__init__.py: Added import statement for IBVPNet
* evaluation/metrics.py: Updated for MACC evaluation metrics
* evaluation/post_process.py: Updated for MACC evaluation metrics
* unsupervised_methods/unsupervised_predictor.py: Updated for MACC evaluation metrics
* main.py: Added support for iBVPNet model
* config.py: Updated for iBVPNet model
* setup.sh: Updated to make the installation work with Ubuntu 20.04 (can be ignored while merging)
* ReadMe: Added details of the iBVP dataset and iBVPNet model

### CODE REVIEW

**Comment by jnj256** [comment]
> I will get to testing and reviewing this PR fully by this weekend @jnj256 - sorry for the wait!

Thanks Akshay for the update. 
I identified few refinements for which I shall raise a separate PR later, as these should not affect your testing. For e.g., at line 139 of metrics.py, there should be no unit for MACC ("dB" needs to be removed).

File: configs/infer_configs/IBVP_PURE_IBVPNET_BASIC.yaml [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+  NAME: iBVPNet
+  IBVPNET:
+    FRAME_NUM: 128
+INFERENCE:
+  BATCH_SIZE: 4
+  EVALUATION_METHOD: "FFT"        # "FFT" or "peak detection"
+  EVALUATION_WINDOW:
+    USE_SMALLER_WINDOW: False        # Change this if you'd like an evaluation window smaller than the test video length
+    WINDOW_SIZE: 10        # In seconds
+  MODEL_PATH: "./final_model_release/IBVP_iBVPNet_diffnormalized.pth"
```
Comment (yahskapar) [inline configs/infer_configs/IBVP_PURE_IBVPNET_BASIC.yaml:48]: Small nitpick, rename the model file itself to have the "i" in "iBVP" consistently lowercase. As in, it should be "iBVP_iBVPNet".

Comment (yahskapar) [reply configs/infer_configs/IBVP_PURE_IBVPNET_BASIC.yaml:48]: If you change this, be sure to change this across all config files after updating the model name in the final_model_release folder.

File: configs/infer_configs/PURE_UNSUPERVISED.yaml [hunk 2]
```diff
@@ -2,14 +2,14 @@ BASE: ['']
 TOOLBOX_MODE: "unsupervised_method"      # "train_and_test"  or "only_test" or "unsupervised_method"
 UNSUPERVISED:
   METHOD: ["ICA", "POS", "CHROM", "GREEN", "LGI", "PBV"]
-  METRICS: ['MAE', 'RMSE', 'MAPE', 'Pearson', 'SNR', 'BA']
+  METRICS: ['MAE', 'RMSE', 'MAPE', 'Pearson', 'SNR', 'BA', 'MACC']
   DATA:
     FS: 30
     DATASET: PURE
-    DO_PREPROCESS: False               # if first time, should be true
+    DO_PREPROCESS: True               # if first time, should be true
```
Comment (yahskapar) [inline configs/infer_configs/PURE_UNSUPERVISED.yaml:9]: Undo this change

File: configs/infer_configs/PURE_UNSUPERVISED.yaml [hunk 3]
```diff
@@ -2,14 +2,14 @@ BASE: ['']
...
     DATA_FORMAT: NDHWC
-    DATA_PATH: "/gscratch/ubicomp/xliu0/data3/mnt/Datasets/PURE/RawData"                     # Raw dataset path, need to be updated
-    CACHED_PATH: "/gscratch/ubicomp/xliu0/data3/mnt/Datasets/rppg_toolbox/PreprocessedData"    # Processed dataset save path, need to be updated
+    DATA_PATH: "data/PURE_Dataset/raw"                     # Raw dataset path, need to be updated
```
Comment (yahskapar) [inline configs/infer_configs/PURE_UNSUPERVISED.yaml:11]: Undo these changes as well

File: configs/infer_configs/UBFC-rPPG_UNSUPERVISED.yaml [hunk 4]
```diff
@@ -6,10 +6,10 @@ UNSUPERVISED:
   DATA:
     FS: 30
     DATASET: UBFC-rPPG
-    DO_PREPROCESS: False               # if first time, should be true
+    DO_PREPROCESS: True               # if first time, should be true
```
Comment (yahskapar) [inline configs/infer_configs/UBFC-rPPG_UNSUPERVISED.yaml:9]: Undo

File: configs/infer_configs/UBFC-rPPG_UNSUPERVISED.yaml [hunk 5]
```diff
@@ -6,10 +6,10 @@ UNSUPERVISED:
...
     DATA_FORMAT: NDHWC
-    DATA_PATH: "/gscratch/ubicomp/xliu0/data3/mnt/Datasets/UBFC-rPPG/RawData"                     # Raw dataset path, need to be updated
-    CACHED_PATH: "/gscratch/ubicomp/xliu0/data3/mnt/Datasets/rppg_toolbox/PreprocessedData"    # Processed dataset save path, need to be updated
+    DATA_PATH: "data/UBFC-rPPG/data"                     # Raw dataset path, need to be updated
```
Comment (yahskapar) [inline configs/infer_configs/UBFC-rPPG_UNSUPERVISED.yaml:11]: Undo

File: evaluation/metrics.py [hunk 6]
```diff
@@ -129,6 +133,10 @@ def calculate_metrics(predictions, labels, config):
                 SNR_FFT = np.mean(SNR_all)
                 standard_error = np.std(SNR_all) / np.sqrt(num_test_samples)
                 print("FFT SNR (FFT Label): {0} +/- {1} (dB)".format(SNR_FFT, standard_error))
+            elif metric == "MACC":
+                MACC_avg = np.mean(MACC_all)
+                standard_error = np.std(MACC_all) / np.sqrt(num_test_samples)
+                print("MACC: {0} +/- {1} (dB)".format(MACC_avg, standard_error))
```
Comment (yahskapar) [inline evaluation/metrics.py:139]: I know you mentioned you'd remove dB in another PR, but maybe consider doing it here alongside the other small config-related changes I suggested.

File: main.py [hunk 7]
```diff
@@ -150,9 +154,11 @@ def unsupervised_method_inference(config, data_loader):
             train_loader = data_loader.BP4DPlusBigSmallLoader.BP4DPlusBigSmallLoader
         elif config.TRAIN.DATA.DATASET == "UBFC-PHYS":
             train_loader = data_loader.UBFCPHYSLoader.UBFCPHYSLoader
+        elif config.TRAIN.DATA.DATASET == "IBVP":
```
Comment (yahskapar) [inline main.py:157]: Consider changing this, and other instances of this elsewhere, from IBVP to iBVP for consistency.

File: setup.sh [hunk 8]
```diff
@@ -1,2 +1,2 @@
 conda remove --name rppg-toolbox --all -y
-conda create -n rppg-toolbox python=3.8 pytorch=1.12.1 torchvision=0.13.1 torchaudio=0.12.1 cudatoolkit=10.2 -c pytorch -q -y
+conda create -n rppg-toolbox python=3.8 pytorch==1.12.1 torchvision==0.13.1 torchaudio==0.12.1 cudatoolkit=11.6 -c pytorch -c conda-forge -q -y
```
Comment (yahskapar) [inline setup.sh:2]: I'm a little hesitant to allow this change to go through - can you remind me why exactly you had to do this in order to get this to work on your version of Ubuntu? FWIW, I also use Ubuntu 20.04 and haven't run into any issues with the original environment creation command.

Comment (yahskapar) [reply setup.sh:2]: Was it something specific to code that you added (e.g., related to the iBVPNet model)?

**Review by yahskapar — CHANGES_REQUESTED** [review]
Hi @jnj256,

Fairly minor comments, and many of the changes I requested should be quick. Should be good to go after addressing those, and I look forward to the additional PR you mentioned with some other changes.

File: setup.sh (hunk 8 above)
Comment (jnj256) [reply setup.sh]: This must have been related to driver versions, which are following for me: 
Driver Version: 530.30.02    CUDA Version: 12.1
And therefore, the corresponding cudatoolkit (cudatoolkit=11.6) was required for installation. 
However, given your comment, I've reverted this change.

**Comment by jnj256** [comment]
@yahskapar : Thank you for going through all the changes made and for your suggestions. I've now addressed your comments and suggestions. Please have a look if everything looks good to go.

**Review by yahskapar — APPROVED** [review]
LGTM, thanks!

**Comment by yahskapar** [comment]
@jnj256, looks like there's some conflicts, probably because of the OMIT PR that was merged recently. Can you pull those changes and make sure to resolve any conflicts? You should be good to squash and merge after that - thanks again for your contribution!

**Comment by jnj256** [comment]
@yahskapar : Thanks for your quick review of the changes. About the conflict related to OMIT PR, I've now resolved it.

**Pull request merged**